### PR TITLE
[ContactStructuralMechanicsApplication] WIP Frictional contact

### PR DESCRIPTION
**Pull request by loumalouomega**
This is WIP for frictional contact. What is updated:

- The local computation of the RHS and LHS have access to the ProcessInfo (constant)
- The main variables related with friction are now in components (instead of resultant as double)
- Many utilities are updated in order to take into account this changes (search, criterion, etc...)
- The frictional formulation has been updated (the new frictional conditions has been created with the AD, that why there are +138000 lines of codes)
- A minimal test has been added. I will add more tests when I solve the problem of memory related with #1468 

Requires #1795 in order to work

### CODE REVIEW

**Comment by RiccardoRossi** [comment]
any hope you can simplify it a bit? 138k lines of code is a little too much...

File: applications/ContactStructuralMechanicsApplication/symbolic_generation/ALM_frictionless_mortar_condition/generate_frictionless_mortar_condition.py [hunk 1]
```diff
@@ -31,12 +31,12 @@ def convert_active_inactive_int(list_active):
...
...atrix<double, TNumNodes, TDim>& X2 = rDerivativeData.X2;\n    \n    const array_1d<double, TNumNodes>& LMNormal = MortarUtilities::GetVariableVector<TNumNodes>(this->GetGeometry(), NORMAL_CONTACT_STRESS, 0);\n    \n    const bounded_matrix<double, TNumNodes, TDim>& NormalSlave = rDerivativeData.NormalSlave;\n\n    // The ALM parameters\n    const array_1d<double, TNumNodes>& DynamicFactor = MortarUtilities::GetVariableVector<TNumNodes>(this->GetGeometry(), DYNAMIC_FACTOR);\n    const double ScaleFactor = rDerivativeData.ScaleFactor;\n    const array_1d<double, TNumNodes>& PenaltyParameter = rDerivativeData.PenaltyParameter;\n    \n    // Mortar operators\n    const bounded_matrix<double, TNumNodes, TNumNodes>& MOperator = rMortarConditionMatrices.MOperator;\n    const bounded_matrix<double, TNumNodes, TNumNodes>& DOperator = rMortarConditionMatrices.DOperator;\n    // Mortar operators derivatives\n    const array_1d<bounded_matrix<double, TNumNodes, TNumNodes>, SIZEDERIVATIVES2>& DeltaMOperator = rMortarConditionMatrices.DeltaMOperator;\n    const array_1d<bounded_matrix<double, TNumNodes, TNumNodes>, SIZEDERIVATIVES2>& DeltaDOperator = rMortarConditionMatrices.DeltaDOperator;\n\n"
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/symbolic_generation/ALM_frictionless_mortar_condition/generate_frictionless_mortar_condition.py]: why don't you leave it as "human readable" without the \n s in the code...

**Comment by loumalouomega** [comment]
The AD is automatic, I can not control the quantity of output lines

File: applications/ContactStructuralMechanicsApplication/symbolic_generation/ALM_frictionless_mortar_condition/generate_frictionless_mortar_condition.py (hunk 1 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/symbolic_generation/ALM_frictionless_mortar_condition/generate_frictionless_mortar_condition.py]: It was giving me problems, that's why I replaced the new lines with \n

**Comment by loumalouomega** [comment]
@RiccardoRossi I solved the conflicts due to the pybind, this branch is ready for merging

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp [hunk 2]
```diff
@@ -80,12 +81,12 @@ void AugmentedLagrangianMethodFrictionlessComponentsMortarContactCondition<2,2,
...
     const bounded_matrix<double, 2, 2>& X2 = rDerivativeData.X2;
     
-    const bounded_matrix<double, 2, 2>& LM = MortarUtilities::GetVariableMatrix<2, 2>(this->GetGeometry(), VECTOR_LAGRANGE_MULTIPLIER, 0);
+    const bounded_matrix<double, 2, 2> LM = MortarUtilities::GetVariableMatrix<2, 2>(this->GetGeometry(), VECTOR_LAGRANGE_MULTIPLIER, 0);
     
     const bounded_matrix<double, 2, 2>& NormalSlave = rDerivativeData.NormalSlave;
 
     // The ALM parameters
-    const array_1d<double, 2>& DynamicFactor = MortarUtilities::GetVariableVector<2>(this->GetGeometry(), DYNAMIC_FACTOR);
+    const array_1d<double, 2> DynamicFactor = MortarUtilities::GetVariableVector<2>(this->GetGeometry(), DYNAMIC_FACTOR);
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp:89]: here the & was correct

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp [hunk 3]
```diff
@@ -42368,12 +42373,12 @@ void AugmentedLagrangianMethodFrictionlessComponentsMortarContactCondition<3,3,
     const bounded_matrix<double, 3, 3>& X1 = rDerivativeData.X1;
     const bounded_matrix<double, 3, 3>& X2 = rDerivativeData.X2;
     
-    const bounded_matrix<double, 3, 3>& LM = MortarUtilities::GetVariableMatrix<3, 3>(this->GetGeometry(), VECTOR_LAGRANGE_MULTIPLIER, 0);
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp:42371]: hrtr yjr & is valid

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp [hunk 4]
```diff
@@ -42368,12 +42373,12 @@ void AugmentedLagrangianMethodFrictionlessComponentsMortarContactCondition<3,3,
...
+    const bounded_matrix<double, 3, 3> LM = MortarUtilities::GetVariableMatrix<3, 3>(this->GetGeometry(), VECTOR_LAGRANGE_MULTIPLIER, 0);
     
     const bounded_matrix<double, 3, 3>& NormalSlave = rDerivativeData.NormalSlave;
 
     // The ALM parameters
-    const array_1d<double, 3>& DynamicFactor = MortarUtilities::GetVariableVector<3>(this->GetGeometry(), DYNAMIC_FACTOR);
+    const array_1d<double, 3> DynamicFactor = MortarUtilities::GetVariableVector<3>(this->GetGeometry(), DYNAMIC_FACTOR);
     const double ScaleFactor = rDerivativeData.ScaleFactor;
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp:42382]: same

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp [hunk 5]
```diff
@@ -51703,12 +51709,12 @@ void AugmentedLagrangianMethodFrictionlessComponentsMortarContactCondition<3,4,
     const bounded_matrix<double, 4, 3>& X1 = rDerivativeData.X1;
     const bounded_matrix<double, 4, 3>& X2 = rDerivativeData.X2;
     
-    const bounded_matrix<double, 4, 3>& LM = MortarUtilities::GetVariableMatrix<3, 4>(this->GetGeometry(), VECTOR_LAGRANGE_MULTIPLIER, 0);
+    const bounded_matrix<double, 4, 3> LM = MortarUtilities::GetVariableMatrix<3, 4>(this->GetGeometry(), VECTOR_LAGRANGE_MULTIPLIER, 0);
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp:51712]: same

File: applications/ContactStructuralMechanicsApplication/tests/ALM_frictional_contact_test_2D/hyper_simple_patch_test.mdpa [hunk 6]
```diff
@@ -0,0 +1,108 @@
+Begin ModelPartData
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/tests/ALM_frictional_contact_test_2D/hyper_simple_patch_test.mdpa:1]: can't you do without mdpa if it is so simple?

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
you took out a lot of &.

this is bad for performance (they were valid!!)

File: applications/ContactStructuralMechanicsApplication/tests/ALM_frictional_contact_test_2D/hyper_simple_patch_test.mdpa (hunk 6 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/tests/ALM_frictional_contact_test_2D/hyper_simple_patch_test.mdpa:1]: But this way is easier...

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp (hunk 2 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp:89]: take into account I am building an auxiliary matrix, is not directly the FastGetSolutionStepValue

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp (hunk 3 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp:42371]: Idem

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp (hunk 4 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp:42382]: Idem

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp (hunk 5 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp:51712]: Idem

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp (hunk 2 above)
Comment (RiccardoRossi) [reply applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp:89]: If you are building an auxiliary matrix, why do you need the GetVariable?

Comment (RiccardoRossi) [reply applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictionless_components_mortar_contact_condition.cpp:89]: well ok, you may need it initialized. anyway...your call

**Review by RiccardoRossi — APPROVED** [review]

**Pull request merged**